File: cvdnn.py
# please refer to: http://www.pyimagesearch.com/2017/08/21/deep-learning-with-opencv/
# for detail

import numpy as np
import argparse as ap
import time
import cv2

# parse command line arguments
arp = ap.ArgumentParser()
arp.add_argument("-i", "--image", required=False,
	help="path to input image")
arp.add_argument("-p", "--prototxt", required=True,
	help="path to Caffe 'deploy' prototxt file")
arp.add_argument("-m", "--model", required=True,
	help="path to Caffe pretrained model")
arp.add_argument("-l", "--labels", required=True,
	help="path to ImageNet labels (i.e., syn-sets)")
args = vars(arp.parse_args())

# Use live demo if image is not given
cap = cv2.VideoCapture()
if (args["image"] is None) :
   cap.open(0)
else :
   cap.open(args["image"])

if (cap.isOpened() == False) :
   print("ERROR: Fail to capture")
   exit(0)

# set opencv video capture properties
cap.set(cv2.CAP_PROP_FRAME_WIDTH, 224)
cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 224)

# load labels
rows = open(args["labels"]).read().strip().split("\n")
classes = [r[r.find(" ") + 1:].split(",")[0] for r in rows]

# load model
print("[INFO] loading model...")
net = cv2.dnn.readNetFromCaffe(args["prototxt"], args["model"])

# loop 
while True:
   
   # read new frame
   ret, image = cap.read()
   if (image is None) : break
   
   # create blob from image
   blob = cv2.dnn.blobFromImage(image, 1, (224, 224), (104, 117, 123))
   
   # assign net input
   net.setInput(blob)
   
   # time before forwarding
   start = time.time()
   
   # forward
   preds = net.forward()
   
   # time after forwarding
   end = time.time()
   
   # show classification time cost
   print("[INFO] classification took {:.5} seconds".format(end - start))

   # find the top 3 scored classes
   idxs = np.argsort(preds[0])[::-1][:3]
   for (i, idx) in enumerate(idxs):
      # mark the top one
	   if i == 0:
		   text = "Label: {}, {:.2f}".format(classes[idx], preds[0][idx])
		   cv2.putText(image, text, (5, 25),  cv2.FONT_HERSHEY_SIMPLEX,
			   0.7, (0, 0, 255), 2)
	   # for each class in top 3, print the info
	   print("[INFO] {}. label: {}, probability: {:.5}".format(i + 1,
		   classes[idx], preds[0][idx]))
   
   # show image
   cv2.imshow("cvdnn_py", image)
   cv2.waitKey(1)

# before exiting
cv2.waitKey(0)
cap.release()
cv2.destroyAllWindows()

# end of file

File: cvdnn.cpp
/**
 * Copyright (c) 2017, Jack Mo (mobangjack@foxmail.com).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <opencv2/opencv.hpp>
#include <opencv2/dnn.hpp>
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>

// command line help
void help()
{
	std::cout << "Usage: cvdnn <image> <proto> <model> <label>" << std::endl;
}

// load labels from file
bool load_labels(const std::string& path, std::vector<std::string>& labels)
{
   std::ifstream label_stream(path.c_str());
   
	if (!label_stream)
		return false;

	while (!label_stream.eof())
	{
		std::string line;
		std::getline(label_stream, line);
		size_t label_beg_idx = line.find(" ") + 1;
		size_t label_end_idx = line.find(",", label_beg_idx + 1);
		if (label_end_idx == std::string::npos) label_end_idx = line.length();
		size_t label_len = label_end_idx - label_beg_idx;
		std::string label = line.substr(label_beg_idx, label_len);
		labels.push_back(label);
	}
	
	label_stream.close();
	
	return true;
}

int main(int argc, char** argv)
{
   if (argc != 5)
   {
	   help();
	   return -1;
   }
	
	// std string for each argument
   std::string image_path(argv[1]);
   std::string proto_path(argv[2]);
   std::string model_path(argv[3]);
   std::string label_path(argv[4]);
	
	// load labels from file
	std::vector<std::string> labels;
	if (!load_labels(label_path, labels))
	{
	   std::cout << "ERROR: cannot open label file: " << label_path << std::endl;
	   return -1;
	}

   // read image
	cv::Mat image = cv::imread(image_path);
	if (image.empty())
	{
	   std::cout << "ERROR: cannot open image file: " << image_path << std::endl;
	   return -1;
	}
	
	// create blob from image
	cv::Mat blob = cv::dnn::blobFromImage(image, 1, cv::Size(224, 224), cv::Scalar(104, 117, 123));

	// read net from caffe
	cv::dnn::Net net = cv::dnn::readNetFromCaffe(proto_path, model_path);
	
	// set net input
	net.setInput(blob);

	// time tick before forwarding
	double t = cv::getTickCount();

	// forward
	cv::Mat probs = net.forward();
	
	// calculate time cost
	t = (cv::getTickCount() - t) / cv::getTickFrequency();

	// show forward time cost
	std::cout << "[INFO] classification took " << (int)t << " seconds" << std::endl;
	
	// debug
	//std::cout << "[INFO] probs.rows: " << probs.rows << ", probs.cols: " << probs.cols << std::endl;

	// sort result
	cv::Mat idxs;
	cv::sortIdx(probs, idxs, cv::SORT_EVERY_ROW + cv::SORT_DESCENDING);

	// debug
	//std::cout << "[INFO] idxs.rows: " << idxs.rows << ", idxs.cols: " << idxs.cols << std::endl;

	// show the top 3 scored classes
	for (int i = 0; i < 3; i++)
	{
		int idx = idxs.at<int>(0, i);
		std::string label = labels[idx];
		float prob = probs.at<float>(0, idx);

		std::stringstream ss;
		ss << label << ": " << prob;
		std::string text = ss.str();

		std::cout << "[INFO] " << (i + 1) << ". " << text << std::endl;
		
		if (i == 0)
			cv::putText(image, text, cv::Size(5, 25), cv::FONT_HERSHEY_SIMPLEX, 0.7, cv::Scalar(0, 0, 255), 2);
	}

	cv::imshow("cvdnn_cpp", image);
	cv::waitKey(0);
	
	cv::destroyAllWindows();

	return 0;
}

// end of file
